entry/src/entry_batch.rs

```diff
@@ -0,0 +1,555 @@
+use {
+    crate::entry::Entry,
+    serde::{
+        de::{self, Visitor},
+        Deserialize, Deserializer, Serialize, Serializer,
+    },
+    solana_clock::Slot,
+    solana_hash::Hash,
+    std::fmt,
+};
+
+/// A batch of entries with optional special metadata.
+///
+/// EntryBatch enforces validation rules but allows flexible combinations:
+/// - Entries can be present with or without special data
+/// - Special data can be present with or without entries
+/// - Only restriction: entries cannot be both present and special data present simultaneously
+/// - Empty entries with no special data is allowed
+///
+/// # Serialization Format
+/// The binary format is:
+/// ```
+/// [8 bytes: entries_len as u64 little-endian]
+/// [variable: serialized entries using bincode, concatenated]
+/// [variable: optional VersionedSpecialEntry if present]
+/// ```
+pub struct EntryBatch {
+    pub entries: Vec<Entry>,
+    pub special: Option<VersionedSpecialEntry>,
```

Comment:
Special entries can't have any entries present `Vec<Entry>` right?
What do you think about making this an enum instead:
- If Vec<Entry>::len() > 0, deserialize as Vec<Entry>
- Else deserialize as a VersionedSpecialEntry